### PR TITLE
dispatchers: allow leading whitespace in window parameter

### LINKED ISSUE #7999: Dispatcher `movewindowpixel` and `resizewindowpixel` fail to match if window parameter includes leading whitespace
### Already reported ? *

- [X] I have searched the existing open and closed issues.

### Regression?

No

### System Info and Version

<details> <summary>System/Version info</summary>

```sh
Hyprland, built from branch main at commit d35e70a8c6599bb058cf86eb87c783ce1cf72471  (cmake: drop ninja dep).
Date: Sat Sep 14 04:56:44 2024
Tag: v0.43.0-22-gd35e70a8, commits: 5218

flags: (if any)


System Information:
System name: Linux
Node name: abyssian-trigrade
Release: 6.10.10-zen1-1-zen
Version: #1 ZEN SMP PREEMPT_DYNAMIC Thu, 12 Sep 2024 17:17:51 +0000


GPU information: 
0a:00.0 VGA compatible controller [0300]: NVIDIA Corporation AD106 [GeForce RTX 4060 Ti] [10de:2803] (rev a1) (prog-if 00 [VGA controller])
NVRM version: NVIDIA UNIX x86_64 Kernel Module  560.35.03  Fri Aug 16 21:39:15 UTC 2024


os-release: NAME="Arch Linux"
PRETTY_NAME="Arch Linux"
ID=arch
BUILD_ID=rolling
ANSI_COLOR="38;2;23;147;209"
HOME_URL="https://archlinux.org/"
DOCUMENTATION_URL="https://wiki.archlinux.org/"
SUPPORT_URL="https://bbs.archlinux.org/"
BUG_REPORT_URL="https://gitlab.archlinux.org/groups/archlinux/-/issues"
PRIVACY_POLICY_URL="https://terms.archlinux.org/docs/privacy-policy/"
LOGO=archlinux-logo


plugins:

======Config-Start======


Config File: /home/ennis/.config/hypr/hyprland.conf: Read Succeeded
# Startup Apps
exec-once = sleep 3 && waybar -l trace |& tee ~/waybar_log
exec-once = systemctl --user start xdg-desktop-portal.service
exec-once = /usr/lib/polkit-kde-authentication-agent-1
exec-once = python3 /home/ennis/.config/hypr/blender_float.py
exec-once = sleep 40 && aw-qt
# Variable Declarations
$terminal = kitty
$fileManager = dolphin
$menu = rofi -show drun

# Env Vars
env = XCURSOR_SIZE,24

# == QT Configuration
env = GDK_BACKEND,wayland,x11,*
env = QT_QPA_PLATFORM,wayland;xcb
env = QT_QPA_PLATFORMTHEME,qt6ct # change to qt6ct if you have that
env = QT_WAYLAND_DISABLE_WINDOWDECORATION,1

# == NVidia Fixups
env = LIBVA_DRIVER_NAME,nvidia
env = GBM_BACKEND,nvidia-drm
env = __GLX_VENDOR_LIBRARY_NAME,nvidia

# == Java Fixering
env = _JAVA_AWT_WM_NONREPARENTING,1

# == XDG shit
env = XDG_SESSION_TYPE,wayland
env = XDG_MENU_PREFIX=arch-
#env = XDG_CURRENT_DESKTOP,Hyprland
#env = XDG_SESSION_DESKTOP,Hyprland

# Monitor Layout
monitor=HDMI-A-1, preferred, 0x0, 1, vrr, 1
monitor=DP-2,preferred,auto,1

# For all categories, see https://wiki.hyprland.org/Configuring/Variables/
input {
    kb_layout = us
    kb_variant =
    kb_model =
    kb_options =
    kb_rules =

    follow_mouse = 1
    mouse_refocus = false
    float_switch_override_focus = 2

    touchpad {
        natural_scroll = no
    }

    sensitivity = 0 # -1.0 - 1.0, 0 means no modification.
}

general {
    # See https://wiki.hyprland.org/Configuring/Variables/ for more

    gaps_in = 2
    gaps_out = 2
    border_size = 2
    col.active_border = rgba(2f4f4fee) rgba(00ff99ee) 45deg
    col.inactive_border = rgba(595959aa)

    layout = dwindle

    # Please see https://wiki.hyprland.org/Configuring/Tearing/ before you turn this on
    allow_tearing = true

    resize_on_border = true
}

cursor {
    no_warps = 1
    no_hardware_cursors = true
    enable_hyprcursor = true
}

decoration {
    # See https://wiki.hyprland.org/Configuring/Variables/ for more

    rounding = 4
    
    blur {
        enabled = false
        size = 3
        passes = 1
    }

    drop_shadow = yes
    shadow_range = 4
    shadow_render_power = 3
    col.shadow = rgba(1a1a1aee)
}

animations {
    animation=windows,0
    animation=workspaces,0
    animation=layers,0
    animation=fade,0
    animation=border,0
}

dwindle {
    # See https://wiki.hyprland.org/Configuring/Dwindle-Layout/ for more
    pseudotile = yes # master switch for pseudotiling. Enabling is bound to mainMod + P in the keybinds section below
    preserve_split = yes # you probably want this
}

master {
    # See https://wiki.hyprland.org/Configuring/Master-Layout/ for more
    #xnew_is_master = true
}

gestures {
    # See https://wiki.hyprland.org/Configuring/Variables/ for more
    workspace_swipe = off
}

misc {
    # See https://wiki.hyprland.org/Configuring/Variables/ for more
    force_default_wallpaper = 1 # Set to 0 or 1 to disable the anime mascot wallpapers
    vrr = 2
}


# Window Rules
windowrulev2 = suppressevent maximize, class:.* 

# render:explicit_sync = no

# -- Quakitty


# -- File Dialogs
windowrulev2 = float,class:^org.freedesktop.impl.portal.*$
windowrulev2 = float,title:(^Open Files$)

# -- Steam
windowrulev2 = float,class:(steam),title:(^Friends List$)
windowrulev2 = float,class:(steam),title:(^Steam - News$)
windowrulev2 = float,class:(steam),title:(^.* - Chat$)
windowrulev2 = float,class:(steam),title:(^* - event started$)
windowrulev2 = float,class:(steam),title:(^.* CD key$)
windowrulev2 = float,class:(steam),title:(^Steam - Self Updater$)
windowrulev2 = float,class:(steam),title:(^Screenshot Uploader$)
windowrulev2 = float,class:(steam),title:(^Steam Guard - Computer Authorization Required$)

# -- Krita
windowrulev2 = minsize 1024 640,class:(krita),title:(^Open Images — Krita$)
windowrulev2 = minsize 1024 640,class:(krita),title:(^Saving As — Krita$)

# -- Blender
windowrulev2 = float,title:^(Blender Preferences)$

# -- Ondsel
windowrulev2 = stayfocused,title:^(Expression editor)$

# -- Dropdown Terminal
windowrulev2 = animation slidevert,class:^quakitty$

# -- Bricscad
windowrulev2 = decorate off,class:(Bricscad),title:(^LookFrom|QUAD$)
windowrulev2 = noborder on,class:(Bricscad),title:(^LookFrom|QUAD$)
windowrulev2 = noshadow on,class:(Bricscad),title:(^LookFrom|QUAD$)
windowrulev2 = noinitialfocus,class:(Bricscad),title:(^LookFrom|QUAD$)

# -- PlugData
windowrulev2 = noinitialfocus,class:(^plugdata$)

# -- Godot
windowrulev2 = minsize 640 480,class:(Godot),title:^(Please confirm...|Resource)$
windowrulev2 = center,class:(Godot),title:^(Please confirm...|Resource)$

# -- Fix JetBrains
#windowrulev2 = noinitialfocus,class:^jetbrains-(?!toolbox),floating:1
#windowrulev2 = stayfocused,class:^(jetbrains-.*)
#windowrulev2 = noinitialfocus, class:^(.*jetbrains.*)$, title:^(win.*)$
#windowrulev2 = nofocus, class:^(.*jetbrains.*)$, title:^(win.*)$
windowrulev2 = noinitialfocus, class:^(jetbrains-.*)$, title:^$

# -- VRRTest
windowrulev2 = immediate on,title:^Freesync Test$

# Workspace Rules
workspace = 1,monitor:HDMI-A-1,default:true,persistent:true
workspace = 2,monitor:HDMI-A-1
workspace = 3,monitor:HDMI-A-1
workspace = 4,monitor:HDMI-A-1
workspace = 5,monitor:HDMI-A-1
workspace = 6,monitor:DP-2,default:true,persistent:true
workspace = 7,monitor:DP-2
workspace = 8,monitor:DP-2
workspace = 9,monitor:DP-2
workspace = 10,monitor:DP-2

# split-monitor-workspaces config

# See https://wiki.hyprland.org/Configuring/Keywords/ for more
$mainMod = SUPER

# Example binds, see https://wiki.hyprland.org/Configuring/Binds/ for more
bind = $mainMod, T, exec, $terminal
bind = $mainMod SHIFT, Q, killactive, 
bind = $mainMod SHIFT, E, exit, 
bind = $mainMod, E, exec, $fileManager
bind = $mainMod, D, exec, $menu
bind = $mainMod, P, pseudo, # dwindle
bind = $mainMod, J, togglesplit, # dwindle
bind = $mainMod SHIFT, Space, togglefloating
bind = $mainMod SHIFT, F, fullscreen

# Move focus with mainMod + arrow keys
bind = $mainMod, left, movefocus, l
bind = $mainMod, right, movefocus, r
bind = $mainMod, up, movefocus, u
bind = $mainMod, down, movefocus, d

# Switch workspaces with mainMod + [0-9]
bind = $mainMod, 1, focusmonitor, 1
# bind = $mainMod, 1, moveworkspacetomonitor,1 1
bind = $mainMod, 1, workspace, 1

bind = $mainMod, 2, focusmonitor, 1
# bind = $mainMod, 2, moveworkspacetomonitor,2 1
bind = $mainMod, 2, workspace, 2

bind = $mainMod, 3, focusmonitor, 1
# bind = $mainMod, 3, moveworkspacetomonitor,3 1
bind = $mainMod, 3, workspace, 3

bind = $mainMod, 4, focusmonitor, 1
# bind = $mainMod, 4, moveworkspacetomonitor,4 1
bind = $mainMod, 4, workspace, 4

bind = $mainMod, 5, focusmonitor, 1
# bind = $mainMod, 5, moveworkspacetomonitor,5 1
bind = $mainMod, 5, workspace, 5

bind = $mainMod, 6, focusmonitor, 0
# bind = $mainMod, 6, moveworkspacetomonitor,6 0
bind = $mainMod, 6, workspace, 6

bind = $mainMod, 7, focusmonitor, 0
# bind = $mainMod, 7, moveworkspacetomonitor,7 0
bind = $mainMod, 7, workspace, 7

bind = $mainMod, 8, focusmonitor, 0
# bind = $mainMod, 8, moveworkspacetomonitor,8 0
bind = $mainMod, 8, workspace, 8

bind = $mainMod, 9, focusmonitor, 0
# bind = $mainMod, 9, moveworkspacetomonitor,9 0
bind = $mainMod, 9, workspace, 9

bind = $mainMod, 0, focusmonitor, 0
# bind = $mainMod, 0, moveworkspacetomonitor,10 0
bind = $mainMod, 0, workspace, 10

# Move active window to a workspace with mainMod + SHIFT + [0-9]
bind = $mainMod SHIFT, 1, movetoworkspacesilent, 1
bind = $mainMod SHIFT, 2, movetoworkspacesilent, 2
bind = $mainMod SHIFT, 3, movetoworkspacesilent, 3
bind = $mainMod SHIFT, 4, movetoworkspacesilent, 4
bind = $mainMod SHIFT, 5, movetoworkspacesilent, 5
bind = $mainMod SHIFT, 6, movetoworkspacesilent, 6
bind = $mainMod SHIFT, 7, movetoworkspacesilent, 7
bind = $mainMod SHIFT, 8, movetoworkspacesilent, 8
bind = $mainMod SHIFT, 9, movetoworkspacesilent, 9
bind = $mainMod SHIFT, 0, movetoworkspacesilent, 10

# Example special workspace (scratchpad)
bind = $mainMod, S, togglespecialworkspace, magic
bind = $mainMod SHIFT, S, movetoworkspace, special:magic

# Scroll through existing workspaces with mainMod + scroll
bind = $mainMod, mouse_down, workspace, e+1
bind = $mainMod, mouse_up, workspace, e-1

# Move/resize windows with mainMod + LMB/RMB and dragging
bindm = $mainMod, mouse:272, movewindow
bindm = $mainMod, mouse:273, resizewindow

# Media Controls
bind=, XF86AudioRaiseVolume, exec, wpctl set-volume -l 1.5 @DEFAULT_AUDIO_SINK@ 1%+
bind=, XF86AudioLowerVolume, exec, wpctl set-volume -l 1.5 @DEFAULT_AUDIO_SINK@ 1%-
bind=, XF86AudioMute, exec, wpctl set-mute @DEFAULT_AUDIO_SINK@ toggle

bind=, XF86AudioPlay, exec, playerctl play-pause
bind=, XF86AudioPrev, exec, playerctl previous
bind=, XF86AudioNext, exec, playerctl next
bind=, code:248, exec, grim -o $(hyprctl -j monitors | jq -r '.[] | select(.focused) | .name') - | wl-copy

# Dropdown Terminal
bind=,F12,exec,~/.config/hypr/hdrop -c quakitty -f -p t -w 75 'kitty --class quakitty'
#debug:disable_logs = false
exec-once = sleep 1.5 && hyprctl dispatch movetoworkspacesilent "1" && hyprctl dispatch movetoworkspacesilent "6"

======Config-End========
```

</details>

### Description

The dispatchers `resizewindowpixel` and `movewindowpixel` will both fail to match windows at all if any leading whitespace is present in the window specifier.

### How to reproduce

Run the following:
- `hyprctl dispatch resizewindowpixel 20 20, class:<CLASS>`, replacing `<CLASS>` with the window class of your choice
- The above will fail, printing `resizeWindow: no window`
- Running `hyprctl dispatch resizewindowpixel 20 20,class:<CLASS>` will work as expected
- This is also true for `movewindowpixel`

This issue also occurs if the commands are bound in the configuration, i.e. `bind = $mainMod, right, movewindowpixel, 20 20, class:Code` will do nothing, but  `bind = $mainMod, right, movewindowpixel, 20 20,class:Code` will work fine

### Crash reports, logs, images, videos

_No response_

### PR DESCRIPTION
**Pull request by JManch**
#### Describe your PR, what does it fix/add?
Fixes #7999

Trims leading whitespace when checking for keywords. Still uses the untrimmed regex if no keyword is provided.

#### Is there anything you want to mention? (unchecked code, possible bugs, found problems, breaking compatibility, etc.)

No

#### Is it ready for merging, or does it need work?

Ready

### CODE REVIEW

File: src/Compositor.cpp [hunk 1]
```diff
@@ -2446,36 +2446,20 @@ void CCompositor::scheduleFrameForMonitor(CMonitor* pMonitor, IOutput::scheduleF
 }
 
 PHLWINDOW CCompositor::getWindowByRegex(const std::string& regexp) {
-    if (regexp.starts_with("active"))
-        return m_pLastWindow.lock();
-
-    eFocusWindowMode mode = MODE_CLASS_REGEX;
+    // Trim leading whitespace
+    auto tRegexp = regexp;
+    auto pos     = tRegexp.find_first_not_of(" \t\n\r");
```
Comment (vaxerski) [inline src/Compositor.cpp:2451]: trim()

File: src/Compositor.cpp [hunk 2]
```diff
@@ -2487,6 +2471,30 @@ PHLWINDOW CCompositor::getWindowByRegex(const std::string& regexp) {
         return nullptr;
     }
 
+    eFocusWindowMode mode = MODE_CLASS_REGEX;
+
+    std::regex       regexCheck(regexp);
+    std::string      matchCheck;
+
+    if (tRegexp.starts_with("class:")) {
```
Comment (vaxerski) [inline src/Compositor.cpp:2479]: why all these changes and shuffling around?

File: src/Compositor.cpp (hunk 1 above)
Comment (JManch) [reply src/Compositor.cpp:2451]: I thought removing trailing whitespace could potentially change the regex

File: src/Compositor.cpp (hunk 2 above)
Comment (JManch) [reply src/Compositor.cpp:2479]: It's cause I grouped the "floating" and "tiled" checks above with "active" as those cases can return early without creating a regex object

File: src/Compositor.cpp (hunk 1 above)
Comment (vaxerski) [reply src/Compositor.cpp:2451]: we already do that in other places so it shouldnt matter

**Review by vaxerski — APPROVED** [review]
thanks!

**Pull request merged**